clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -225,11 +230,20 @@ public double compressionRate() {
     }
 
     public boolean isFull() {
-        return recordsBuilder.isFull();
+        return !isWritable || recordsBuilder.isFull();
```

Comment:
So, looking over it a bit more, I now know why I introduced `isWritable`. The `isFull` method is not only used during append. It is also used to wake the sender up: if the deque is of size one and the only batch in there is not full, the sender will not be woken up to drain until it hits the linger ms. By introducing `isWritable` we can get the batch to be drained slightly quicker.

Not sure if the extra state is worth that optimization though.